File: algorithms/tree/design/fenwick_tree.py
# Fenwick Tree (Binary Indexed Tree)
#
# Time complexity:
# Building a Fenwick tree:  O(n log n)
# Optimal construction:     O(n)
# Updating an element:      O(log n)
# Prefix sum:               O(log n)
# Sum over a range:         O(log n)
#
# Memory usage:             O(n)

class FenwickTree:
    def __init__(self, n):
        self.n = n
        self.bit = [0] * (n + 1)

    # Add function (also known as update),
    # and ofcourse you can use it like -value for subtract from bit, but I think it's not that cool
    def add(self, i, value):
        while i <= self.n:
            self.bit[i] += value
            i += i & -i

    def subtract(self, i, value):
        while i <= self.n:
            self.bit[i] -= value
            i += i & -i

    def prefix_sum(self, i):
        prefix_sum = 0

        while i > 0:
            prefix_sum += self.bit[i]
            i -= i & -i

        return prefix_sum

    def range_sum(self, start, end):
        return self.prefix_sum(end) - self.prefix_sum(start - 1)

# How to use:
#
# # creating
# ft = FenwickTree(n
#
# for i in range(1, n + 1):
#     ft.add(i, a[i])
#
# # update
# delta = new_value - a[i]
# a[i] = new_value
# ft.add(i, delta)
#
# # query
# res = ft.range_sum(l, r)
# print(res)
#
# res = ft.prefix_sum(x)

# Intuition:
# Fenwick Tree это способ хранить суммы перекрывающихся диапазонов через битовые прыжки
# Fenwick Tree это структура, которая позволяет собирать и разбирать суммы через прыжки по младшему установленному биту числа
#
# Главная проблема:
# Быстро поддерживать массив, где нужно:
#
# менять один элемент
# быстро считать сумму на префиксе / отрезке
#
# Обычный массив:
# update: O(1)
# prefix sum: O(n) ❌
#
# Prefix sums массив:
# update: O(n) ❌
# query: O(1)
#
# Fenwick Tree — компромисс:
# update: O(log n)
# query: O(log n)

# Idea:
# Fenwick Tree хранит суммы блоков разного размера, причём эти блоки определяются через:
# i & -i
#
# Это выражение даёт:
# младший установленный бит числа i
# (LSB — least significant bit)

# Where to use:
# 1. Много запросов “изменить + посчитать сумму”
# 2. Частоты / подсчёт количества элементов
# 3. Инверсии (очень важная классика)
#           Сколько пар (i < j), где a[i] > a[j]
# 4. “Сколько элементов в диапазоне значений”
# 5. нельзя заранее обработать все запросы

File: cpp_algorithms/tree/design/fenwick_tree.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

class FenwickTree {
private:
    int n;
    vector<ll> bit;

public:
    explicit FenwickTree(int size) : n(size), bit(size + 1, 0) {}

    explicit FenwickTree(const vector<int>& values)
        : n(static_cast<int>(values.size())), bit(n + 1, 0) {

        for (int i = 1; i <= n; i++) {
            bit[i] += values[i - 1];

            int parent = i + (i & -i);

            if (parent <= n) {
                bit[parent] += bit[i];
            }
        }
    }

    void add(int i, ll value) {
        while (i <= n) {
            bit[i] += value;
            i += i & -i;
        }
    }

    void subtract(int i, ll value) {
        add(i, -value);
    }

    ll prefix_sum(int i) {
        ll res = 0;

        while (i > 0) {
            res += bit[i];
            i -= i & -i;
        }

        return res;
    }

    ll range_sum(int left, int right) {
        return prefix_sum(right) - prefix_sum(left - 1);
    }
};

// Usage example:

// #include <bits/stdc++.h>
// using namespace std;
// using ll = long long;
//
// int main() {
//     FenwickTree fenwick(5);

//     fenwick.add(1, 2);
//     fenwick.add(2, 4);
//     fenwick.add(3, 7);
//     fenwick.add(4, 3);
//     fenwick.add(5, 5);

//     cout << fenwick.prefix_sum(3) << '\n';  // 2 + 4 + 7 = 13
//     cout << fenwick.range_sum(2, 4) << '\n'; // 4 + 7 + 3 = 14

//     fenwick.subtract(3, 2); // the value in position 3 has decreased from 7 to 5

//     cout << fenwick.range_sum(2, 4) << '\n'; // 4 + 5 + 3 = 12
// }
